internal/controller/deployment_controller.go

```diff
@@ -409,55 +435,58 @@ func (r *DeploymentReconciler) Delete(ctx context.Context, l logr.Logger, deploy
 	return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
 }
 
-func (r *DeploymentReconciler) releaseAWSCluster(ctx context.Context, namespace, clusterName, templateName string) error {
-	aws, err := r.hasAWSProvider(ctx, templateName)
+func (r *DeploymentReconciler) releaseCluster(ctx context.Context, namespace, clusterName, templateName string) error {
+	prividers, err := r.getProviders(ctx, templateName)
 	if err != nil {
 		return err
 	}
-	if !aws {
-		// nothing to do - we can skip this step
-		return nil
-	}
+	if slices.Contains(prividers, "aws") {
```

Comment:
The only thing that differs between these two blocks in the provider name string and the `gvk` being used by `removeClusterFinalizer`, the more providers we add the longer this will grow.

:golf: 

```
type providerSchema struct {
	machine, cluster schema.GroupVersionKind
}

func (r *DeploymentReconciler) releaseCluster(ctx context.Context, namespace, clusterName, templateName string) error {
	providers, err := r.getProviders(ctx, templateName)
	if err != nil {
		return err
	}

	providerGVKs := map[string]providerSchema{
		"aws":   providerSchema{machine: gvkAWSMachine, cluster: gvkAWSCluster},
		"azure": providerSchema{machine: gvkAzureMachine, cluster: gvkAzureCluster},
	}

	// Associate the provider with it's GVK
	for _, provider := range providers {
		gvk := providerGVKs[provider]

		found, err := r.machinesAvailable(ctx, namespace, clusterName, gvk.machine)
		if err != nil {
			return err
		}

		if !found {
			return r.removeClusterFinalizer(ctx, namespace, clusterName, gvk.cluster)
		}

	}

	return nil
}
```

Reply:
Thanks for the improvement